pkg/controllers/disruption/orchestration/queue.go

```diff
@@ -0,0 +1,326 @@
+		q.recorder.Publish(disruptionevents.Launching(nodeClaim, cmd.Reason()))
+		initializedStatus := nodeClaim.StatusConditions().GetCondition(v1beta1.Initialized)
+		if !initializedStatus.IsTrue() {
+			q.recorder.Publish(disruptionevents.WaitingOnReadiness(nodeClaim))
+			waitErrs[i] = fmt.Errorf("node claim not initialized")
+			continue
+		}
+		cmd.ReplacementKeys[i].Initialized = true
+		// Subtract the last initialization time from the time the command was added to get initialization duration.
+		initLength := initializedStatus.LastTransitionTime.Inner.Time.Sub(nodeClaim.CreationTimestamp.Time).Seconds()
+		disruptionReplacementNodeClaimInitializedHistogram.Observe(initLength)
+	}
+	// If we have any errors, don't continue
+	if err := multierr.Combine(waitErrs...); err != nil {
+		return fmt.Errorf("waiting for replacement initialization, %w", err)
+	}
+
+	// All replacements have been provisioned.
+	// All we need to do now is get a successful delete call for each node claim,
+	// then the termination controller will handle the eventual deletion of the nodes.
+	var multiErr error
+	for i := range cmd.Candidates {
+		candidate := cmd.Candidates[i]
+		q.recorder.Publish(disruptionevents.Terminating(candidate.Node, candidate.NodeClaim, cmd.Reason())...)
+		if err := nodeclaim.Delete(ctx, q.kubeClient, candidate.NodeClaim); err != nil {
+			multiErr = multierr.Append(multiErr, client.IgnoreNotFound(err))
+		} else {
+			nodeclaim.TerminatedCounter(cmd.Candidates[i].NodeClaim, cmd.Method).Inc()
+		}
+	}
+	// If there were any deletion failures, we should requeue.
+	// In the case where we requeue, but the timeout for the command is reached, we'll mark this as a failure.
+	if multiErr != nil {
+		return fmt.Errorf("terminating nodeclaims, %w", multiErr)
+	}
+	return nil
+}
+
+// Add adds commands to the Queue
+// Each command added to the queue should already be validated and ready for execution.
+func (q *Queue) Add(cmd *Command) error {
+	providerIDs := lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string {
+		return s.ProviderID()
+	})
+	// First check if we can add the command.
+	if q.HasAny(providerIDs...) {
+		return fmt.Errorf("candidate is being disrupted")
+	}
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	for _, candidate := range cmd.Candidates {
+		q.providerIDToCommand[candidate.ProviderID()] = cmd
+	}
+	q.RateLimitingInterface.Add(cmd)
+	return nil
+}
+
+// HasAny checks to see if the candidate is part of an currently executing command.
+func (q *Queue) HasAny(ids ...string) bool {
+	q.mu.RLock()
+	defer q.mu.RUnlock()
+
+	// If the mapping has at least one of the candidates' providerIDs, return true.
+	_, ok := lo.Find(ids, func(id string) bool {
+		_, ok := q.providerIDToCommand[id]
+		return ok
+	})
+	return ok
+}
+
+// Remove fully clears the queue of all references of a hash/command
+func (q *Queue) Remove(cmd *Command) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	q.RateLimitingInterface.Done(cmd)
```

Comment:
nit: Should this be next to the `Forget` call just so since the calls are related?